### ISSUE TITLE
Feature: Real ID.

### ISSUE THREAD

**Issue by heyitsbench** [azerothcore/azerothcore-wotlk#14764]
### Describe your feature request or suggestion in detail

Real ID was implemented in patch 3.3.5. From what I can gather, it did seem to be an in-client solution that allowed players to communicate across all characters of an account, as well as messaging across Wow and SC2 at the time. The way a friend could be added with real ID was by going into the social menu and adding the friend by their account's e-mail address rather than their character's username. Apparently the recipient of the friend request will have it show up as pending and will have to accept it for the two to be registered as friends. I'm not sure if this was handled by Battle.net or what. Can't find any opcodes specific to this feature, but I'm not very good with ctrl+f.

**Comment by AnchyDev** [comment]
This was handled by battle.net I'm pretty sure so you would also need to create a battle.net server.

**Comment by Kitzunu** [comment]
Afaik Battlenet server was introduced in Cata, and so was RealId?

**Comment by AnchyDev** [comment]
RealId was implemented in 3.3.5, before 3.3.5a but the features dont show on the client unless it creates a successful connection to the battle net server.

**Comment by noisiver** [comment]
The obvious tell is that you can input an email-address as the username and it'll try to connect using battle.net.

If battle.net were to be implemented, I doubt it though, RealID would be the last thing on peoples mind at that point.

**Comment by heyitsbench** [comment]
My personal main interest with this would be adding friends by their account, which is a real ID function. In my quick exploration, I figured that part would be something that *wasn't* handled by Battle.net, but perhaps I was wrong. 🤷

**Comment by heyitsbench** [comment]
It doesn't seem to be the client asking Battle.net for the account with the e-mail address, as the worldserver *does* get the `CMSG_ADD_FRIEND` opcode:
```
C->S: [Player: Druna GUID Full: 0x000000000000000e Type: Player Low: 14, Account: 1] [CMSG_ADD_FRIEND 0x0069 (105)]
WORLD: Received CMSG_ADD_FRIEND
WORLD: Druna asked to add friend : 'heyitsbench@fake_e-mail_for_github_issue_but_i_did_use_the_e-mail_in_the_account_table_when_testing.co.uk.com.net.gov.edu'
```
It just doesn't do anything with it. Obviously not sure how the server should be giving any information on a real ID friend.

**Comment by noisiver** [comment]
Are you sure it's not the usual add friend that you have in the game right now?

**Comment by AnchyDev** [comment]
This is the frame logic when adding a friend, so a different OPCode could be being sent when BattleNet is connected.
```lua
function AddFriendFrame_Accept()
	local name = AddFriendNameEditBox:GetText();
	if ( AddFriendFrame.BNconnected and string.find(name, "@") ) then
		BNSendFriendInvite(name, AddFriendNoteEditBox:GetText());
	else
		AddFriend(name);
	end
	StaticPopupSpecial_Hide(AddFriendFrame);
end
```
or it could be communicating a different way entirely but I am unsure.

**Comment by AnchyDev** [comment]
Try running `BNSendFriendInvite` and see if you receive anything on the server.

**Comment by heyitsbench** [comment]
> Try running BNSendFriendInvite and see if you receive anything on the server.

C->S: [Player: Druna GUID Full: 0x000000000000000e Type: Player Low: 14, Account: 1] [CMSG_ADD_FRIEND 0x0069 (105)]
WORLD: Received CMSG_ADD_FRIEND
WORLD: Druna asked to add friend : 'fake_e-mail_for_github@azerothcore.org'

**Comment by AnchyDev** [comment]
> > Try running BNSendFriendInvite and see if you receive anything on the server.
> 
> ```
> C->S: [Player: Druna GUID Full: 0x000000000000000e Type: Player Low: 14, Account: 1] [CMSG_ADD_FRIEND 0x0069 (105)]
> WORLD: Received CMSG_ADD_FRIEND
> WORLD: Druna asked to add friend : 'fake_e-mail_for_github@azerothcore.org'
> ```

I just booted up my server and I'm not getting anything on the server? What did you run exactly?

**Comment by heyitsbench** [comment]
> I just booted up my server and I'm not getting anything on the server? What did you run exactly?

Replaced L1307 with `BNSendFriendInvite(name, "");` and L2638 with `BNSendFriendInvite(name, "");` in FriendsFrame.lua.

**Comment by AnchyDev** [comment]
That's odd because I am using the actual real id frame and it's not getting anything on the server, are you sure it patched correctly?

**Comment by heyitsbench** [comment]
> are you sure it patched correctly?

I have no reason to suspect it isn't. 🤷

**Comment by AnchyDev** [comment]
I am receiving the correct opcodes when entering a name without `@` included in the name, however when entering an email in the frame its not receiving anything. My assumption is that it's trying to communicate with battlenet through the battle.net dll but there is no connection so the function never succeeds. I have no way to prove this however.

**Comment by noisiver** [comment]
> I am receiving the correct opcodes when entering a name without `@` included in the name, however when entering an email in the frame its not receiving anything. My assumption is that it's trying to communicate with battlenet through the battle.net dll but there is no connection so the function never succeeds. I have no way to prove this however. EDIT: This is with the realid frame enabled of course, not the normal add friend frame.

I think that's right. I have no way to prove it either but RealID **is** a part of Battle.net.

I honestly think what Bench is triggering is the usual add friends opcode and not Real ID.

*As of Patch 3.3.5, World of Warcraft players have access to a brand-new feature called Real ID, a completely voluntary and optional level of identity that will keep players connected across all of Battle.net®.*

**Comment by AnchyDev** [comment]
This can probably be closed because I think it is outside of the scope of the core.

**Comment by heyitsbench** [comment]
I disagree. The core aims to emulate 3.3.5a, and real ID is in 3.3.5a. But of course my opinion means jack, so do what you want.

**Comment by AnchyDev** [comment]
> I disagree. The core aims to emulate 3.3.5a, and real ID is in 3.3.5a. But of course my opinion means jack, so do what you want.

What I am referring to is I think realid is mostly handled by the client which we cannot change. Your opinion does not mean jack! 😄

**Comment by noisiver** [comment]
It's not really *mostly* handled by the client. The friends you add to real id aren't removed if you fully reset your client. It was also used to speak to friends cross-realm and either back then or later on across different games.

**Comment by AnchyDev** [comment]
But this communication layer would be through battle.net dll would it not?

**Comment by noisiver** [comment]
Most likely
